clang/unittests/Format/NumericLiteralCaseTest.cpp

```diff
@@ -0,0 +1,354 @@
+  // Floating point literals without suffixes.
+  std::vector<StringRef> FloatingPointStatements = {
+      StringRef("a = 0."),          StringRef("b = 1.0"),
+      StringRef("c = .123'45E-10"), StringRef("d = 12'3.0e1"),
+      StringRef("e = 0Xa0eE.P10"),  StringRef("f = 0xeE01.aFf3p6"),
+  };
+
+  struct FloatSuffix {
+    StringRef Lower;
+    StringRef Upper;
+  };
+  // All legal floating point literals defined in the C++23 standard.
+  std::vector<FloatSuffix> FloatingPointSuffixes = {
+      {"f", "F"},     {"l", "L"},       {"f16", "F16"},   {"f32", "F32"},
+      {"f64", "F64"}, {"f128", "F128"}, {"bf16", "BF16"},
+  };
+
+  // Test all combinations of literals with suffixes.
+  for (auto &Statement : FloatingPointStatements) {
+    for (auto &Suffix : FloatingPointSuffixes) {
+      std::string LowerLine =
+          std::string{Statement} + std::string{Suffix.Lower} + ";";
+      std::string UpperLine =
+          std::string{Statement} + std::string{Suffix.Upper} + ";";
+
+      Style.NumericLiteralCase.Suffix = FormatStyle::NLCS_Leave;
+      verifyFormat(LowerLine, Style);
+      verifyFormat(UpperLine, Style);
+
+      Style.NumericLiteralCase.Suffix = FormatStyle::NLCS_Lower;
+      verifyFormat(LowerLine, Style);
+      verifyFormat(LowerLine, UpperLine, Style);
+
+      Style.NumericLiteralCase.Suffix = FormatStyle::NLCS_Upper;
+      verifyFormat(UpperLine, LowerLine, Style);
+      verifyFormat(UpperLine, Style);
+    }
+  }
+}
+
+TEST_F(NumericLiteralCaseTest, CppStandardAndUserDefinedLiteralsAreUntouched) {
+  FormatStyle Style = getLLVMStyle();
+  Style.NumericLiteralCase.Prefix = FormatStyle::NLCS_Upper;
+  Style.NumericLiteralCase.HexDigit = FormatStyle::NLCS_Upper;
+  Style.NumericLiteralCase.ExponentLetter = FormatStyle::NLCS_Upper;
+  Style.NumericLiteralCase.Suffix = FormatStyle::NLCS_Upper;
+
+  // C++ user-defined suffixes begin with '_' or are reserved for the standard
+  // library.
+  const StringRef UDLiterals{"a = 12.if;\n"
+                             "b = -3i;\n"
+                             "c = 100'01il;\n"
+                             "d = 100'0.12il;\n"
+                             "e = 12h;\n"
+                             "f = 0XABE12h;\n"
+                             "g = 0XFA03min;\n"
+                             "h = 0X12B4Ds;\n"
+                             "i = 20.13E-1ms;\n"
+                             "j = 20.13E-1us;\n"
+                             "k = 20.13E-1ns;\n"
+                             "l = 20.13E-1y;\n"
+                             "m = 20.13E-1d;\n"
+                             "n = 20.13E-1d;\n"
+                             "o = 1d;\n"
+                             "p = 102_ffl_lzlz;\n"
+                             "q = 10.2_l;\n"
+                             "r = 0XABDE.0'1P-23_f;\n"
+                             "s = 102_foo_bar;\n"
+                             "t = 123.456_felfz_ballpen;\n"
+                             "u = 0XBEAD1_spacebar;\n"};
```

Comment:
```suggestion
                             "u = 0XBEAD1_spacebar;"};
```
and elsewhere. Except in rare cases, we don't end test cases with a newline.